loopd/daemon.go

```diff
@@ -119,117 +135,291 @@ func daemon(config *config, lisCfg *listenerCfg) error {
 	// We'll also create and start an accompanying proxy to serve clients
 	// through REST.
 	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
+	d.restCtxCancel = cancel
 	mux := proxy.NewServeMux(customMarshalerOption)
 	var restHandler http.Handler = mux
-	if config.CORSOrigin != "" {
-		restHandler = allowCORS(restHandler, config.CORSOrigin)
+	if d.cfg.CORSOrigin != "" {
+		restHandler = allowCORS(restHandler, d.cfg.CORSOrigin)
 	}
 	proxyOpts := []grpc.DialOption{
 		grpc.WithInsecure(),
 		grpc.WithDefaultCallOptions(maxMsgRecvSize),
 	}
 	err = looprpc.RegisterSwapClientHandlerFromEndpoint(
-		ctx, mux, config.RPCListen, proxyOpts,
+		ctx, mux, d.cfg.RPCListen, proxyOpts,
 	)
 	if err != nil {
-		return err
+		return d.stopAndReturnErr(err)
 	}
 
-	restListener, err := lisCfg.restListener()
+	d.restListener, err = d.listenerCfg.restListener()
 	if err != nil {
-		return fmt.Errorf("REST proxy unable to listen on %s",
-			config.RESTListen)
+		return d.stopAndReturnErr(fmt.Errorf("REST proxy unable to "+
+			"listen on %s: %v", d.cfg.RESTListen, err))
 	}
 
 	// A nil listener indicates REST is disabled.
-	if restListener != nil {
+	if d.restListener != nil {
 		log.Infof("Starting REST proxy listener")
 
-		defer restListener.Close()
-		proxy := &http.Server{Handler: restHandler}
+		d.restServer = &http.Server{Handler: restHandler}
 
+		d.wg.Add(1)
 		go func() {
-			err := proxy.Serve(restListener)
+			defer d.wg.Done()
+
+			log.Infof("REST proxy listening on %s",
+				d.restListener.Addr())
+			err := d.restServer.Serve(d.restListener)
 			// ErrServerClosed is always returned when the proxy is
 			// shut down, so don't log it.
 			if err != nil && err != http.ErrServerClosed {
 				log.Error(err)
+				d.errChan <- err
 			}
 		}()
 	} else {
 		log.Infof("REST proxy disabled")
 	}
 
-	mainCtx, cancel := context.WithCancel(context.Background())
-	var wg sync.WaitGroup
+	// Start the grpc server.
+	d.wg.Add(1)
+	go func() {
+		defer d.wg.Done()
+
+		log.Infof("RPC server listening on %s", d.grpcListener.Addr())
+		err = d.grpcServer.Serve(d.grpcListener)
+		if err != nil && err != grpc.ErrServerStopped {
+			log.Error(err)
+			d.errChan <- err
+		}
+	}()
+
+	return nil
+}
+
+// StartAsSubserver is an alternative to Start where the RPC server does not
+// create its own gRPC server but registers to an existing one. The same goes
+// for REST (if enabled), instead of creating an own mux and HTTP server, we
+// register to an existing one.
+func (d *Daemon) StartAsSubserver(lndGrpc *lndclient.GrpcLndServices) error {
```

Comment:
So this function for now ignores the err channel?

Reply:
No. You still have to instantiate the daemon with the channel and react on an error arriving there. But the `d.stopAndReturnErr` was missing here.

Reply:
Ah right, same `NewDaemon` call.